### PR TITLE
mv: add progress bar

### PR DESCRIPTION
**Pull request by ctsk**
Similarly to `cp`, adds `-g` and `--progress` flags to enable a progress bar via indicatif.

The logic is tricker than the progress bar logic in `cp` and modeled closely to  the implementation in advcpmv. In particular, the progress bar does not calculate the total size of the files if the move is performed via a rename operation. To still show adequate feedback, we use two progress bars:

1. Count-Progress-Bar
  - This progress bar indicates how many of the source arguments have been moved. 
  - Only shows up if there's 2 or more source arguments

2. Folder-Copy-Progress-Bar
  - This progress bar indicates how much of a folder has been moved.
  - Only shows up if the source is a folder and copy+remove is required. Doesn't show up for file copies, as there's no progress reporting for that.

<details>
<summary>Screenshots</summary>

Multiple a file and a folder are mv'd across filesystems:

Screenshot from 2022-12-07 14-47-03

Screenshot from 2022-12-07 14-47-07

</details>

### CODE REVIEW

File: src/uu/mv/src/mv.rs [hunk 1]
```diff
@@ -441,7 +509,46 @@ fn rename_with_fallback(from: &Path, to: &Path) -> io::Result<()> {
...
+                None
+            };
+
+            let progress_handler = |process_info: TransitProcess| {
+                if let Some(ref pb) = progress_bar {
+                    pb.set_position(process_info.copied_bytes);
+                    pb.set_message(process_info.file_name);
+                }
+                TransitProcessResult::ContinueOrAbort
+            };
```
Comment (tertsdiepraam) [inline src/uu/mv/src/mv.rs:549]: I think I prefer something like this to make it a bit simpler in my opinion:
```rust
let result = if let Some(ref pb) = progress_bar {
    move_dir_with_progress(from, to, &options, |process_info| {
        pb.set_position(process_info.copied_bytes);
        pb.set_message(process_info.file_name);
        TransitProcessResult::ContinueOrAbort
    })
} else {
    move_dir(from, to, &options)
}
```

File: src/uu/mv/src/mv.rs [hunk 2]
```diff
@@ -441,7 +509,46 @@ fn rename_with_fallback(from: &Path, to: &Path) -> io::Result<()> {
...
+
+            // Calculate total size of directory
+            // Silently degrades:
+            //    If finding the total size fails for whatever reason,
+            //    the progress bar wont be shown for this file / dir.
+            //    (Move will probably fail due to permission error later?)
+            let total_size = match dir_get_size(from) {
+                Ok(size) => Some(size),
+                Err(_) => None,
+            };
```
Comment (tertsdiepraam) [inline src/uu/mv/src/mv.rs:521]:
```suggestion
            let total_size = dir_get_size(from).ok();
```

File: src/uu/mv/src/mv.rs [hunk 3]
```diff
@@ -441,7 +509,46 @@ fn rename_with_fallback(from: &Path, to: &Path) -> io::Result<()> {
...
+                        ProgressStyle::with_template(
+                            "{msg}: [{elapsed_precise}] {wide_bar} {bytes:>7}/{total_bytes:7}",
+                        )
+                        .unwrap(),
+                    );
+
+                    match multi_progress {
+                        Some(mp) => Some(mp.add(bar)),
+                        None => Some(bar),
+                    }
```
Comment (tertsdiepraam) [inline src/uu/mv/src/mv.rs:535]: I think we can simplify this logic, by always making a `MultiProgress` if `b.progress_bar` is `true`. Just with either zero or one bar already in it. Then we only have to match on `multi_progress` and check `b.progress_bar` anymore.

File: src/uu/mv/src/mv.rs [hunk 4]
```diff
@@ -24,9 +25,12 @@ use std::path::{Path, PathBuf};
 use uucore::backup_control::{self, BackupMode};
 use uucore::display::Quotable;
 use uucore::error::{FromIo, UError, UResult, USimpleError, UUsageError};
-use uucore::{format_usage, prompt_yes, show, show_if_err};
+use uucore::{format_usage, prompt_yes, show};
 
-use fs_extra::dir::{move_dir, CopyOptions as DirCopyOptions};
+use fs_extra::dir::{
+    get_size as dir_get_size, move_dir_with_progress, CopyOptions as DirCopyOptions,
```
Comment (tertsdiepraam) [inline src/uu/mv/src/mv.rs:31]: Cool! We might want to look into using that same `get_size` function in `cp` as well (in another PR).

**Review by tertsdiepraam — CHANGES_REQUESTED** [review]
Excellent! I really like this design. My only important suggestion is that we can simplify the logic around the `MultiProgress`, by always initializing one if the progress bar flag is passed. Then we can always just add to it instead of having to branch between a `MultiProgress` and a single `ProgressBar`.

It's a bit unfortunate that we have to take `suspend` into account for all the prints, but I don't have a solution for that :)

**Comment by ctsk** [comment]
Thank you for all the feedback!

**Review by tertsdiepraam — APPROVED** [review]
Looks good! Thanks!

**Pull request merged**